deepCR/training.py

```diff
@@ -0,0 +1,217 @@
+""" module for training new deepCR-mask models
+"""
+import torch
+from tqdm import tqdm as tqdm
+from deepCR.unet import WrappedModel, UNet2Sigmoid
+import matplotlib.pyplot as plt
+import torch.nn as nn
+import torch.optim as optim
+from torch.autograd import Variable
+from torch.utils.data import DataLoader
+import numpy as np
+import datetime
+from torch.optim.lr_scheduler import ReduceLROnPlateau
+from deepCR.util import maskMetric
+from deepCR.dataset import dataset
+
+__all__ = 'train'
+
+class train():
+
+    def __init__(self, image, mask, ignore=None, sky=None, name='model', hidden=32, gpu=False, epoch=50, batch_size=16, lr=0.005, aug_sky=[0, 0], save_after=0, plot_every=10, directory='./'):
+        """ This is the class for training deepCR-mask.
+        :param image: np.ndarray (N*W*W) training data: image array with CR.
+        :param mask: np.ndarray (N*W*W) training data: CR mask array
+        :param ignore: training data: Mask for taking loss. e.g., bad pixel, saturation, etc.
+        :param name: model name. model saved to name_epoch.pth
+        :param hidden: number of channels for the first convolution layer. default: 50
+        :param gpu: True if use GPU for training
+        :param epoch: Number of epochs to train. default: 50
+        :param batch_size: training batch size. default: 16
+        :param lr: learning rate. default: 0.005
+        :param aug_sky: [float, float]. Use random sky background from aug_sky[0] * sky to aug_sky[1] * sky.
+        :param save_after: epoch after which trainer automatically saves model state with lowest validation loss
+        :param plot_every: for every epoch, visualize mask prediction for 1st image in validation set.
+        """
+        if sky is None and aug_sky != [0, 0]:
+            raise AttributeError('Var (sky) is required for sky background augmentation!')
+        if ignore is None:
+            ignore = np.zeros_like(image)
+        assert image.shape == mask.shape == ignore.shape
+        assert image.shape[1] == image.shape[2]
+        data_train = dataset(image, mask, ignore, sky, part='train', aug_sky=aug_sky)
+        data_val = dataset(image, mask, ignore, sky, part='val', aug_sky=aug_sky)
+        self.TrainLoader = DataLoader(data_train, batch_size=batch_size, shuffle=True, num_workers=1)
+        self.ValLoader = DataLoader(data_val, batch_size=batch_size, shuffle=False, num_workers=1)
+        self.shape = image.shape[1]
+        self.name = name
+
+        if gpu:
+            self.dtype = torch.cuda.FloatTensor
+            self.dint = torch.cuda.ByteTensor
+            self.network = nn.DataParallel(UNet2Sigmoid(1,1,hidden))
+            self.network.type(self.dtype)
+        else:
+            self.dtype = torch.FloatTensor
+            self.dint = torch.ByteTensor
+            self.network = WrappedModel(UNet2Sigmoid(1,1,hidden))
+            self.network.type(self.dtype)
+
+        self.optimizer = optim.Adam(self.network.parameters(), lr=lr)
```

Comment:
consider making `optimizer` and `lr_scheduler` as keywords supplied  by the user, with sensible defaults.